[DS-4167] Migrate update-sequences.sql to `database` command

terrywbrady opened the pull request:
Fixes #7509

Migrate update-sequences.sql to the _dspace database_ command.

In addition to being convenient, this will enable dspace docker instances populated from AIP content to have their sequences automatically updated.

```
$ docker exec dspace //dspace/bin/dspace database update-sequences
Running org/dspace/storage/rdbms/sqlmigration/postgres/update-sequences.sql
update-sequences complete
```

Since the dspace/etc/ directory will be going away, move the SQL to dspace-api/src/main/resources to reside with the Flyway SQL.

### Migration needed
- 6x
- 5x
- 4x (maybe)

### Input needed
- Should this PR remove dspace/etc directory?
- How can the existing Oracle update-sequences.sql be converted to a single SQL file?  The current solution requires the SQL to reside on the file system.

terrywbrady committed:
Migrate postgres update-sequences.sql

tdonohue commented on dspace-api/src/main/java/org/dspace/storage/rdbms/DatabaseUtils.java:
```diff
@@ -328,24 +330,54 @@ public static void main(String[] argv) {
+                            for (String line = br.readLine(); line != null; line = br.readLine()) {
+                                sb.append(line);
+                            }
+                        }
```
You should be able to replace all this fancy string parsing with just one line:
`String sqlString = IOUtils.toString(sqlstream, "UTF-8");`

That line will convert the UTF-8 InputStream into a String.

IOUtils requires an `import org.apache.commons.io.IOUtils`.

terrywbrady replied:
Added commit 22cc5469a23

tdonohue commented:
@terrywbrady : I gave this a quick look today, and it looks like a good start. I added one inline comment/suggestion.

Regarding your questions in the description:
* I think we should remove the `dspace/etc` directory. After this change it only contains a README.  That README was only there to note that the SQL scripts moved between 4.x and 5.x, and it's already duplicated by documentation (on the wiki).
* You likely won't be able to port this to 4.x, as the `dspace database` commandline tool didn't exist until 5.x. Honestly, I'm a little hesitant to port this back to prior releases, as we don't tend to move around files like this during minor upgrades.  If you need it ported to 5.x/6.x for Docker, maybe we consider leaving the `dspace/etc` folder in 5.x/6.x (and leaving the update-sequence.sql file in that location), so that we don't confuse 5.x/6.x users after their next minor upgrade.  I like this PR though as-is for the 7.x version.
* I don't have an answer on the Oracle question -- hopefully someone with more Oracle experience could give some tips.

mwoodiupui commented:
You may want to look at making `updateseq.sql` into a PL/SQL subprogram, tucking its declaration into `update-sequences.sql`.
BTW the internal documentation for updateseq seems to be incomplete, as it doesn't describe the fourth parameter used with the `handle_seq` table.

mwoodiupui commented on dspace-api/src/main/java/org/dspace/storage/rdbms/DatabaseUtils.java:
```diff
@@ -328,24 +330,54 @@ public static void main(String[] argv) {
+            } else if (argv[0].equalsIgnoreCase("update-sequences")) {
+                try (Connection connection = dataSource.getConnection()) {
+                    String dbType = getDbType(connection);
+                    String sqlfile = "org/dspace/storage/rdbms/sqlmigration/" + dbType +
```
Should this not be either absolute "/org/dspace/storage/rdbms/sqlmigration/" or relative "sqlmigration/"?

terrywbrady replied:
The other worked, but it seems more clear to start with /.  Added commit e421a7c7c75

terrywbrady replied:
The resource is not found if it is prefixed with /.  Perhaps that is treated as a filesystem path.  Reverting to "org/..."

terrywbrady committed:
apply review feedback

terrywbrady committed:
remove /etc directory

terrywbrady committed:
simplify string extract from stream

terrywbrady commented:
@mwoodiupui, I added your recommended Oracle stored procedure.  I haven't yet figured out if I will be able to test this in Docker.  Hopefully we can find someone who can verify the change in a DSpace Oracle DB.

@tdonohue , I have removed the etc directory.  Your recommendations on the port to 5x and 6x make sense.  The process will need to read from the filesystem rather than the classpath in order to execute then change.  I will likely leave the Oracle SQL alone for 5x/6x.  I will tackle that after this code is merged.

terrywbrady committed:
attempt to fix headers

terrywbrady committed:
remove dspace/etc from build steps

terrywbrady committed:
remove /org

terrywbrady committed:
update sql

mwoodiupui committed:
[DS-4167] Fix various problems with Oracle script.

terrywbrady committed:
v1 oracle sql

terrywbrady committed:
Merge branch 'ds4167' into DS-4167

terrywbrady committed:
Merge pull request #15 from mwoodiupui/DS-4167

[DS-4167] Fix various problems with Oracle script.

terrywbrady commented:
@mwoodiupui , your SQL is working as a part of the script.
```
 MINGW64 ~/DSpace-Docker-Images/docker-compose-files/dspace-oracle-compose (master)
$ docker ps
CONTAINER ID        IMAGE                       COMMAND                  CREATED             STATUS              PORTS                              NAMES
c80cee226a93        dspace/dspace:local-build   "catalina.sh run"        25 seconds ago      Up 7 seconds        8009/tcp, 0.0.0.0:8080->8080/tcp   dspace
8aff80560ae9        atmire/dspace-oracle:11g2   "/entrypoint.sh /bin…"   26 seconds ago      Up 25 seconds       1521/tcp, 8080/tcp                 dspacedb

~/DSpace-Docker-Images/docker-compose-files/dspace-oracle-compose (master)
$ docker exec dspace //dspace/bin/dspace database update-sequences
Running org/dspace/storage/rdbms/sqlmigration/oracle/update-sequences.sql
update-sequences complete
```

terrywbrady commented:
@mwoodiupui , @tdonohue , could you give this a quick re-review?  I think I addressed your comments.

tdonohue reviewed (approved):
👍 The code looks fine to me. However, I haven't had a chance to test this on Postgres or Oracle.

mwoodiupui reviewed (approved):
Looks right, works on Oracle, works on PostgreSQL.
